clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -407,10 +407,11 @@ private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestamp
             if (future.succeeded()) {
                 ListOffsetResult value = future.value();
                 result.fetchedOffsets.putAll(value.fetchedOffsets);
-                if (value.partitionsToRetry.isEmpty())
-                    return result;
 
                 remainingToSearch.keySet().removeAll(result.fetchedOffsets.keySet());
+                remainingToSearch.keySet().removeAll(value.partitionsWithUnknownOffset);
+                if (value.partitionsToRetry.isEmpty() && remainingToSearch.isEmpty())
```

Comment:
There seems to be some redundance between `partitionsToRetry` and `remainingToSearch`. It might be nicer if we could get rid of `remainingToSearch` so that we only had to rely on `ListOffsetResult` to know if we should retry. I think the only thing we need is to avoid losing partitions in the call to `groupListOffsetRequests`.

Reply:
Ah yes, we could add partitions we lose in the `groupListOffsetRequest` to `partitionsToRetry`, since `sendListOffsetsRequests` is called only from that one method...  Let me try that.